Revert "Documentation: Remove warning that nonstandard primitive type sizes may reveal LLVM bugs"

LilithHafner opened the pull request:
Reverts JuliaLang/julia#58262

See #58434, for an example bug.

LilithHafner committed:
Revert "Documentation: Remove warning that nonstandard primitive type sizes m…"

This reverts commit 3fc2a159aeeb295dd28e4814b177be6d538e8aba.

LilithHafner committed:
Update types.md

Per https://github.com/JuliaLang/julia/issues/58434#issuecomment-2887208822, seems like this might not be an LLVM bug.

Seelengrab commented:
#58434 is about a multiple of 8 though, how does that issue prove issues with non-multiples of 8?

PatrickHaecker commented:
From reading through the analysis done in #58434 and the PR in #58435, I don't think, that fully reverting #58262 is the correct thing to do.

This is the summary I got from these discussions (with some interpretation from me included):
- A Julia bug was identified, which manifests when using a global `Tuple`/`struct` of a non-power-of-2 `primitive type` and using the `===` comparison (but not the `==` comparison)
- The bug is inside Julia, as we didn't indicate that a data structures needs and uses padding when using a non-power-of-2 primitive type which then lead to a sometimes incorrect comparison with `===` (depending on the values of the padded bits which were accidentally compared, but not set).
- In my understanding not using padding would also be a violation of the C API that we want to follow by default.
- The bug should be fixed with #58435
- This fix should be backported to 1.10, 1.11 and 1.12
- It's plausible, that the bug was not identified by, e.g. `BitIntegers.jl`'s test cases, because they do neither cover globals nor composite types nor tuples and comparison with these types is probably more often done with `==` than with `===`.
- There is still no indication that `LLVM` has bugs in this area.

The fact that LLVM had bugs in the past in an area where we now also have/had a bug, does not seem to logically conclude that LLVM therefore must be buggy.

We might consider adding a generic warning, that "using nonstandard primitive type sizes may reveal bugs" (note the absence of `LLVM` in that phrase), but let's also stick to the data here. We now had one bug in this area in which time frame? Years?
If we'd put a warning into the documentation for some years whenever we had a single bug in an area, we'd need to document a lot.
I'd prefer to only document it if there is an accumulation of bugs or if they cannot be fixed in a limited amount of time.

PatrickHaecker commented:
> #58434 is about a multiple of 8 though, how does that issue show issues with non-multiples of 8? The bug is also unrelated to the exact size, since it's about padding when placed within a struct.

I guess this is a rhetorical question, but you made me curios:
We can't test this, as currently [no such primitive types can be created](https://github.com/JuliaLang/julia/blob/fc456bdd04d6d391381554e5c8e5d9a1e3ca6ce3/src/builtins.c#L2052). In the absence of additional bugs, I'd expect the very same behavior: Without the fix from #58435 we should see the same problems in structs/tuples. With the fix it should work.

And indeed, when you deactivate the last part of this check with the fix on top of Julia master, this works correctly (repeatedly):
```julia
julia> module Foo
                  primitive type UInt33 33 end
                  a = reinterpret(UInt33, ntuple(_ -> 0x0, 5))
                  res = (a,) === (a,)
                  f(x) = (x,) === (x,)
                  res2 = f(a)
                  g() = (a,) === (a,)
                  res3 = g()
                  function h()
                      x = reinterpret(UInt33, ntuple(_ -> 0x0, 5))
                      return (x,) === (x,)
                  end
                  res4 = h()
              end; using .Foo; (Foo.res, Foo.res2, Foo.res3, Foo.res4)
(true, true, true, true)
```
However, I am undecided on whether the `reinterpret` is a bug or a feature.
On the one hand it seems to be the only way to create such an object and given how you would map this type to real hardware constraints, it's probably (haven't checked) more or less guaranteed by LLVM that this should work.
On the other hand, this looks horribly wrong.

But before we decided on how to construct these objects, all methods besides `==` and `===` seem kind of mood.

Seelengrab commented:
> On the one hand it seems to be the only way to create such an object

You can do `Core.zext_int(UInt33, 0x0)`, which would be the correct way of doing that. The `reinterpret` there is technically not ok, since you're moving data into padding (in the best case, in the worst case you're just dropping allocated memory on the floor) and roundtripping back to that tuple type would read that padding (or uninitialized memory), which would be UB.

> On the other hand, this looks horribly wrong.

Nah, having non-multiple-of-8-bits primitive types is perfectly fine & valid. For example, [zig has them](https://ziglang.org/documentation/master/#Integers):

> Zig supports arbitrary bit-width integers, referenced by using an identifier of `i` or `u` followed by digits. For example, the identifier `i7` refers to a signed 7-bit integer. The maximum allowed bit-width of an integer type is `65535`. 

Zig used to use/has used LLVM as a backend, making use of the same functionality Julia uses for primitive types. I think that's a good argument for that feature of LLVM being pretty stable & correct by now.

PatrickHaecker commented:
> > On the other hand, this looks horribly wrong.
> 
> Nah, having non-multiple-of-8-bits primitive types is perfectly fine & valid.

Sorry, this was ambiguous. I was referring to the point that interpreting a 40 bits type as a 33 bits type look horribly wrong (which you confirmed). I agree, that non-multiple-of-8-bits primitive types are fine in general (when Julia supports them properly).

> You can do `Core.zext_int(UInt33, 0x0)`, which would be the correct way of doing that.

Thanks for this tip. However, it does not work in all cases:
```julia
julia> Core.zext_int(UInt7, 0x0)
ERROR: ZExt: output bitsize must be > input bitsize
```

LilithHafner committed:
Merge branch 'master' into revert-58262-patch-2

LilithHafner commented:
The original PR was based on a false premise (that sizes other than those listed above work just fine and haven't had bugs in a while) and should therefore be reverted.

In general, the bar for reverting a mistake is lower than the bar for making most other changes. I don't see any valid objections to reverting in this thread so I'm going to go ahead and merge unless anyone brings up something new.

Notably,

> https://github.com/JuliaLang/julia/issues/58434 is about a multiple of 8 though...

That is irrelevant. Please read the diff of this PR and the PR it reverts and the discussion on the PR it reverts.

> ...I don't think, that fully reverting https://github.com/JuliaLang/julia/pull/58262 is the correct thing to do...

That's quite reasonable. When a PR is partially, but not fully correct, however, it is reasonable to revert it and subsequently, if necessary, try to reland a fixed version. There is no need to block reverting the mistake based on the potential for a fixed version. After this merges, feel free to write more nuanced documentation about the potential bug profile of unconventional primitive sizes.

LilithHafner commented:
Stop bringing up sizes not being divisible by 8. Neither this PR nor the one it reverts change anything about what we say about sizes not divisible by 8. See https://github.com/JuliaLang/julia/pull/58262#pullrequestreview-2803144621 for clarification.

Seelengrab commented:
I usually consider myself a Bayesian. My prior for this working as it should is just a lot higher, because I maintain a library ([FieldFlags.jl](https://github.com/Seelengrab/FieldFlags.jl) - nightly fails due to JET) that fundamentally relies  on this feature. Apart from the linked issue about padding within structs, I have not encountered issues with this. So if we go back to the previous docs, I'd like to know what exactly we think is broken here so that we can fix it & my package stays in working condition.

LilithHafner commented:
We had a warning saying "watch out about nonstandard sizes, they could be broken" and then we removed it. Very soon after, a bug came up about nonstandard sizes. I think we removed the warning too soon and should put it back for a bit.

I don't care who, if anyone, we blame. I'd be fine with "...to experience Julia or LLVM bugs..." => "...to experience bugs...".

LilithHafner commented:
I'm fine with the former.

PatrickHaecker reviewed (commented):
Proposal based on the agreement in the discussion.

LilithHafner committed:
Update doc/src/manual/types.md

Co-authored-by: PatrickHaecker <152268010+PatrickHaecker@users.noreply.github.com>

PatrickHaecker reviewed (approved):
LGTM.
Thank you for listening to diverse opinions and sorry for the extra work due to my original PR.
